In [ ]:
# 1. Fix the Version Conflict (Downgrade Pandas)
!pip install "pandas==2.2.2"

# 2. Re-install LlamaIndex Experimental just to be safe
!pip install llama-index-experimental -U -q

print("Fix Complete.")
print("CRITICAL: Go to 'Runtime' -> 'Restart Session' immediately after this finishes!")

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.5/88.5 kB 4.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 47.6/47.6 kB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.9/11.9 MB 83.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 92.0/92.0 kB 6.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 442.8/442.8 kB 26.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 59.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 191.3/191.3 kB 15.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 303.3/303.3 kB 22.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 160.3/160.3 kB 6.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.9/50.9 kB 3.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 150.7/150.7 kB 9.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 214.0/214.0 kB 15.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 

In [ ]:
# Force uninstall conflicting libraries
!pip uninstall -y pandas llama-index llama-index-core llama-index-experimental

# Install the exact compatible versions
!pip install "pandas==2.2.2" llama-index llama-index-experimental llama-index-llms-gemini llama-index-embeddings-huggingface nest_asyncio -q

print(" Libraries Installed.")
print(" CRITICAL: Now go to the top menu, click 'Runtime' -> 'Restart session'.")
print("If you skip the restart, it will fail again.")

Found existing installation: pandas 2.2.2
Uninstalling pandas-2.2.2:
  Successfully uninstalled pandas-2.2.2
Found existing installation: llama-index-core 0.14.10
Uninstalling llama-index-core-0.14.10:
  Successfully uninstalled llama-index-core-0.14.10
Found existing installation: llama-index-experimental 0.6.3
Uninstalling llama-index-experimental-0.6.3:
  Successfully uninstalled llama-index-experimental-0.6.3
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.7/12.7 MB 73.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 303.3/303.3 kB 19.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.8/51.8 kB 3.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.5/4.5 MB 81.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.9/63.9 kB 4.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 328.2/328.2 kB 21.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.0/88.0 kB 6.9 MB/s eta 0:00:00
✅ Libraries Installed.
⚠️ CRITI

In [ ]:
# 1. Install Google AI SDK & Pandas (Stable & Simple)
!pip install -q -U google-generativeai pandas

import os
import re
import pandas as pd
import google.generativeai as genai
from google.api_core.exceptions import ResourceExhausted


#  PASTE YOUR API KEY HERE
os.environ["GOOGLE_API_KEY"] = "   "

if os.environ["GOOGLE_API_KEY"].startswith("YOUR"):
    print(" STOP: You forgot to paste your API Key!")
else:
    genai.configure(api_key=os.environ["GOOGLE_API_KEY"])

# ==========================================
# 3. DATA PARSING (Using Pandas!)
# ==========================================
def load_data(filename):
    print(f" Parsing {filename} using Pandas...")
    if not os.path.exists(filename):
        print(" Error: File not found. Please upload 'All_Trains_Cleaned.txt'")
        return pd.DataFrame(), {}

    with open(filename, 'r', encoding='utf-8') as f:
        content = f.read()

    # Split by Train Blocks
    blocks = content.split('<TRAIN_START>')
    schedule_rows = []
    train_descriptions = {} # Acts as our "Librarian" knowledge base

    for block in blocks:
        if "TRAIN NO:" not in block: continue

        # Clean up
        clean_block = block.split('<TRAIN_END>')[0].strip()

        # Extract Meta
        t_num = re.search(r"TRAIN NO:\s*(\d+)", clean_block)
        t_name = re.search(r"TRAIN NAME:\s*(.*)", clean_block)

        if t_num:
            no = t_num.group(1)
            name = t_name.group(1).strip() if t_name else "Unknown"

            # Store full text for description search
            train_descriptions[no] = f"Train {no} ({name}). Schedule:\n{clean_block}"

            # Parse Table for Structured Search
            lines = clean_block.split('\n')
            for line in lines:
                if '|' in line and 'STATION NAME' not in line:
                    parts = [p.strip() for p in line.split('|')]
                    if len(parts) >= 4:
                        schedule_rows.append({
                            "train_no": no,
                            "train_name": name,
                            "station": parts[1], # Station Name
                            "arrival": parts[2],
                            "departure": parts[3]
                        })

    # create the DataFrame
    df = pd.DataFrame(schedule_rows)
    print(f" Loaded {len(df)} schedule rows into Pandas.")
    return df, train_descriptions

# Load Data
df, descriptions = load_data("/content/All_Trains_Cleaned.txt")

# ==========================================
# 4. DEFINE TOOLS (The Brains)
# ==========================================

def find_train_route(source: str, destination: str):
    """
    Finds direct trains between a source station and a destination station.
    Args:
        source: The name of the starting station (e.g. 'Thane').
        destination: The name of the ending station (e.g. 'Lasalgaon').
    """
    print(f"\n⚙️ TOOL: Searching Route {source} -> {destination}...")

    # 1. PANDAS FILTERING
    # Find all rows where station matches Source
    src_matches = df[df['station'].str.contains(source, case=False, na=False)]

    # Find all rows where station matches Destination
    dst_matches = df[df['station'].str.contains(destination, case=False, na=False)]

    # 2. LOGIC (Set Intersection)
    # Find train numbers that exist in BOTH lists
    common_train_ids = set(src_matches['train_no']).intersection(set(dst_matches['train_no']))

    if not common_train_ids:
        return f"No direct trains found between {source} and {destination}."

    results = []
    # Loop through the matches to get times
    for t_id in list(common_train_ids)[:10]: # Limit to top 10
        t_name = df[df['train_no'] == t_id].iloc[0]['train_name']

        # Get departure time from Source
        dep_time = src_matches[src_matches['train_no'] == t_id].iloc[0]['departure']

        # Get arrival time at Destination
        arr_time = dst_matches[dst_matches['train_no'] == t_id].iloc[0]['arrival']

        results.append(f"- Train {t_id} ({t_name}): Departs {source} at {dep_time}, Arrives {destination} at {arr_time}")

    return "\n".join(results)

def get_train_details(train_no: str):
    """
    Gets the full schedule and details of a specific train by its number.
    Args:
        train_no: The number of the train (e.g. '11061').
    """
    print(f"\n TOOL: looking up details for {train_no}...")
    return descriptions.get(train_no, "Train details not found in database.")

# ==========================================
# 5. INITIALIZE GEMINI AGENT
# ==========================================
print("\n Initializing Gemini Agent (Native Mode)...")

# Define the tools list
tools_list = [find_train_route, get_train_details]

# Configure the model with tools
model = genai.GenerativeModel(
    model_name='gemini-2.5-flash-lite',
    tools=tools_list,
    system_instruction="""
    You are a Railway Assistant.
    - If the user asks about travel between two points, call 'find_train_route'.
    - If the user asks for details about a specific train number, call 'get_train_details'.
    - Always answer politely based on the tool output.
    """
)

# Start Chat with Automatic Function Calling
chat = model.start_chat(enable_automatic_function_calling=True)

print(" System Ready! (Pandas Powered)")
print("------------------------------------------------")

# ==========================================
# 6. TEST IT
# ==========================================
# ==========================================
# 6. INTERACTIVE CHAT LOOP
# ==========================================
print("\n💬 Chat Session Started! (Type 'exit' or 'quit' to stop)")
print("-------------------------------------------------------")

while True:
    try:
        # 1. Get User Input
        user_input = input("\nYou: ")

        # 2. Check for Exit
        if user_input.lower() in ["exit", "quit", "bye"]:
            print("👋 Exiting chat. Safe travels!")
            break

        # 3. Send to Gemini (It remembers history!)
        response = chat.send_message(user_input)

        # 4. Print Answer
        print(f"AI: {response.text}")

    except Exception as e:
        print(f"⚠️ Error: {e}")

🚀 Parsing /content/All_Trains_Cleaned.txt using Pandas...
✅ Loaded 104076 schedule rows into Pandas.

🤖 Initializing Gemini Agent (Native Mode)...
✅ System Ready! (Pandas Powered)
------------------------------------------------

💬 Chat Session Started! (Type 'exit' or 'quit' to stop)
-------------------------------------------------------

You: I want to travel from Thane to Lasalgaon

⚙️ TOOL: Searching Route Thane -> Lasalgaon...
AI: Here are the trains that run from Thane to Lasalgaon:
- Train 51196 (Ballarshah Wardha Mumbai Passenger): Departs Thane at 11:10:00, Arrives Lasalgaon at 06:53:00
- Train 17618 (Tapovan Express): Departs Thane at 21:15:00, Arrives Lasalgaon at 17:19:00
- Train 51154 (Bhusaval Mumbai Passenger): Departs Thane at 19:15:00, Arrives Lasalgaon at 11:44:00
- Train 12139 (Sevagram Express): Departs Thane at 15:35:00, Arrives Lasalgaon at 19:18:00
- Train 13201 (RAJENDRA NAGAR PATNA - MUMBAI LTT Janta Exp): Departs Thane at 10:55:00, Arrives Lasalgaon at 05:44:

KeyboardInterrupt: Interrupted by user

In [ ]:

!pip install -q -U google-generativeai pandas

import os
import re
import pandas as pd
import google.generativeai as genai
import sys

#  PASTE YOUR API KEY HERE
os.environ["GOOGLE_API_KEY"] = " "

if os.environ["GOOGLE_API_KEY"].startswith("YOUR"):
    print(" STOP: You forgot to paste your API Key in line 14!")
    sys.exit()

genai.configure(api_key=os.environ["GOOGLE_API_KEY"])

# ==========================================
# 2. DATA LOADING (Parses your Text File)
# ==========================================
def load_data(filename):
    print(f" Parsing {filename}...")
    if not os.path.exists(filename):
        print("Error: 'All_Trains_Cleaned.txt' not found. Please upload it!")
        return None, None

    with open(filename, 'r', encoding='utf-8') as f:
        content = f.read()

    # Split by Train Blocks
    blocks = content.split('<TRAIN_START>')
    schedule_rows = []
    train_descriptions = {}

    for block in blocks:
        if "TRAIN NO:" not in block: continue

        # Clean up
        clean_block = block.split('<TRAIN_END>')[0].strip()

        # Extract Meta
        t_num = re.search(r"TRAIN NO:\s*(\d+)", clean_block)
        t_name = re.search(r"TRAIN NAME:\s*(.*)", clean_block)

        if t_num:
            no = t_num.group(1)
            name = t_name.group(1).strip() if t_name else "Unknown"

            # Save full text for "Schedule" lookups
            train_descriptions[no] = f"Train {no} ({name}). Full Schedule:\n{clean_block}"

            # Save rows for "Search" lookups
            lines = clean_block.split('\n')
            for line in lines:
                if '|' in line and 'STATION NAME' not in line:
                    parts = [p.strip() for p in line.split('|')]
                    if len(parts) >= 4:
                        schedule_rows.append({
                            "train_no": no,
                            "train_name": name,
                            "station": parts[1],
                            "arrival": parts[2],
                            "departure": parts[3]
                        })

    df = pd.DataFrame(schedule_rows)
    print(f" Database Ready: {len(df)} rows loaded.")
    return df, train_descriptions

# Run the Loader
df, descriptions = load_data("All_Trains_Cleaned.txt")

if df is None: sys.exit()

# ==========================================
# 3. DEFINE THE 4 UNIVERSAL TOOLS
# ==========================================

def find_route(source: str, destination: str):
    """
    Finds trains running between two specific stations.
    Example: "Thane to Lasalgaon", "How to go from Mumbai to Delhi?"
    """
    print(f"\n TOOL: Connecting {source} -> {destination}...")

    # Filter Source & Destination
    src_df = df[df['station'].str.contains(source, case=False, na=False)]
    dst_df = df[df['station'].str.contains(destination, case=False, na=False)]

    # Find Intersection (Trains that touch both)
    common_ids = set(src_df['train_no']).intersection(set(dst_df['train_no']))

    if not common_ids:
        return f"No direct trains found between {source} and {destination}."

    results = []
    # Loop through common trains to get timings
    for t_id in list(common_ids)[:15]:
        t_name = df[df['train_no'] == t_id].iloc[0]['train_name']
        dep = src_df[src_df['train_no'] == t_id].iloc[0]['departure']
        arr = dst_df[dst_df['train_no'] == t_id].iloc[0]['arrival']
        results.append(f"- {t_id} {t_name} | Dep {source}: {dep} -> Arr {destination}: {arr}")

    return "\n".join(results)

def get_station_trains(station_name: str):
    """
    Lists all trains that stop at a specific station.
    Example: "Trains from Ranchi", "Trains arriving at Ujjain"
    """
    print(f"\n TOOL: Looking up station '{station_name}'...")
    matches = df[df['station'].str.contains(station_name, case=False, na=False)]

    if matches.empty:
        return f"No station found matching '{station_name}'."

    unique_trains = matches[['train_no', 'train_name', 'arrival', 'departure']].drop_duplicates()
    count = len(unique_trains)

    results = [f"Found {count} trains stopping at {station_name}:"]
    for _, row in unique_trains.head(20).iterrows():
        results.append(f"- {row['train_no']} {row['train_name']} (Arr: {row['arrival']} / Dep: {row['departure']})")

    if count > 20:
        results.append(f"...and {count - 20} more.")

    return "\n".join(results)

def search_train_by_name(keyword: str):
    """
    Searches for a train by its name.
    Example: "What is the number of Rajdhani?", "Search for Geetanjali Express"
    """
    print(f"\n TOOL: Searching train name '{keyword}'...")
    matches = df[df['train_name'].str.contains(keyword, case=False, na=False)]
    unique = matches[['train_no', 'train_name']].drop_duplicates().head(10)

    if unique.empty:
        return f"No trains found with name containing '{keyword}'."

    return unique.to_string(index=False)

def get_schedule(train_no: str):
    """
    Get full schedule/route of a specific train number.
    Example: "Route of 12123", "Timetable of train 11061"
    """
    print(f"\n TOOL: Fetching schedule for {train_no}...")
    return descriptions.get(train_no, "Train details not found.")

# ==========================================
# 4. INITIALIZE AGENT
# ==========================================
print("\n Initializing Universal Railway Agent...")

tools_list = [find_route, get_station_trains, search_train_by_name, get_schedule]

model = genai.GenerativeModel(
    model_name='gemini-2.5-flash-lite',
    tools=tools_list,
    system_instruction="""
    You are a Universal Railway Assistant.
    - If user asks A to B? Use 'find_route'.
    - If user asks about one station (from/to/at)? Use 'get_station_trains'.
    - If user searches a name (e.g. Rajdhani)? Use 'search_train_by_name'.
    - If user gives a number? Use 'get_schedule'.
    - Always be helpful and infer the right tool.
    """
)

chat = model.start_chat(enable_automatic_function_calling=True)
print("Agent Ready! (Type 'exit' to quit)")
print("------------------------------------------------")

# ==========================================
# 5. CHAT LOOP
# ==========================================
while True:
    try:
        q = input("\nYou: ")
        if q.lower() in ["exit", "quit", "bye"]:
            print(" Bye!")
            break

        response = chat.send_message(q)
        print(f"AI: {response.text}")
    except Exception as e:
        print(f" Error: {e}")

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 91.2/91.2 kB 3.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.4/12.4 MB 103.0 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
llama-index-readers-file 0.5.5 requires pandas<2.3.0, but you have pandas 2.3.3 which is incompatible.
llama-index-experimental 0.6.3 requires pandas<2.3.0, but you have pandas 2.3.3 which is incompatible.
google-colab 1.0.0 requires pandas==2.2.2, but you have pandas 2.3.3 which is incompatible.
🚀 Parsing All_Trains_Cleaned.txt...
✅ Database Ready: 104076 rows loaded.

🤖 Initializing Universal Railway Agent...
✅ Agent Ready! (Type 'exit' to quit)
------------------------------------------------

You: Find trains from CST to Pune.

🔎 TOOL: Connecting CST -> Pune...
AI: Here are the trains running from CST to Pune:
- 16339 MUMBAI CST - NAGERCOIL Exp
- 51030 

KeyboardInterrupt: Interrupted by user

In [ ]:
You: Which train from Mumbai to Delhi reaches the earliest in the morning

 TOOL: Connecting Mumbai -> Delhi...
AI: The train that reaches the earliest in the morning is the 12471 Mumbai Bandra (T.) - Jammu Tawi SF Swaraj Express, arriving in Delhi at 04:11:00.

You: Are there any 'Rajdhani' trains that go from Mumbai to Delhi

 TOOL: Searching train name 'Rajdhani Mumbai to Delhi'...
AI: I couldn't find any train with 'Rajdhani Mumbai to Delhi' in its name. However, I did find a train that goes from Mumbai to Delhi and is a 'Rajdhani' express: the 12953 MumbaiCentral-H Nizamuddin August Kranti Rajdhani Express.

You: 12951 is also from mumbai to delhi rajdhani express

 TOOL: Fetching schedule for 12951...
AI: Yes, the Mumbai Central-New Delhi Rajdhani Express (12951) runs from Mumbai to Delhi. It departs from Mumbai Central at 16:40:00 and arrives at New Delhi at 08:30:00 on day 2.

You: give all rajdhani trains from Mumbai to Delhi

 TOOL: Connecting Mumbai -> Delhi...
AI: The Rajdhani trains that run from Mumbai to Delhi are:
- 12953 MumbaiCentral-H Nizamuddin August Kranti Rajdhani Express, departing Mumbai at 17:40:00 and arriving in Delhi at 10:55:00.
- 12951 Mumbai Central-New Delhi Rajdhani Express, departing Mumbai at 16:40:00 and arriving in Delhi at 08:30:00.

You: Get the schedule for train 12139 and tell me how many stops it has

 TOOL: Fetching schedule for 12139...
AI: The schedule for train 12139 (Sewagram Express Slip) is as follows:

It departs from Mumbai CST at 15:00:00 and arrives at Nagpur at 06:10:00 on day 2.

This train has 33 stops.

You: Does the 'Punjab Mail' stop at Dadar

 TOOL: Searching train name 'Punjab Mail'...

 TOOL: Fetching schedule for 12138...
AI: The Punjab Mail (train 12138) stops at Dadar Central at 07:13:00 on day 3.

You: List all trains departing from Pune. Give me only the top 5

TOOL: Looking up station 'Pune'...
AI: Here are the top 5 trains departing from Pune:

1. 51451 Pune-Daund- Baramati Passenger (Dep: 07:10:00)
2. 51409 Pune -Shree Chatrapati Shahu M(T) Kolhapur Pass (Dep: 09:15:00)
3. 51418 Pune Karjat Passenger (Dep: 10:35:00)
4. 11014 Coimbatore Mumbai LTT Express (Dep: 10:45:00)
5. 12581 Pune Bhubaneshwar Express (Dep: 11:15:00)

You: "I want to go to a station named 'Varanasi'. What trains go there from Mumbai?

TOOL: Connecting Mumbai -> Varanasi...
AI: Here are the trains that go from Mumbai to Varanasi:

- **11093 MUMBAI CST - VARANASI Mahanagari Exp**: Departs Mumbai at 00:10:00 and arrives in Varanasi at 04:40:00.
- **11094 VARANASI - MUMBAI CST Mahanagari Exp**: Departs Mumbai at 13:50:00 and arrives in Varanasi at an unspecified time.

You: "Search for 'Duronto' trains. Do any of them stop at Nagpur

 TOOL: Searching train name 'Duronto'...
AI: Yes, the Mumbai CSTM-Nagpur Duronto Express (train 12289) stops at Nagpur.

You: What is the train number for the Deccan Queen?
 TOOL: Searching train name 'Deccan Queen'...
AI: The train number for the Deccan Queen is 12123 and 12124.

ou: How do I go from Thane to New York?
AI: I can only provide information about train routes. New York is not a valid train station. Would you like me to find a train route between two other stations?

You: Find trains from Thane to Lasalgaon and tell me which one takes the least amount of time?

🔎 TOOL: Connecting Thane -> Lasalgaon...
AI: The fastest train from Thane to Lasalgaon is the 17058 Devagiri Express, departing Thane at 06:20:00 and arriving at Lasalgaon at 02:18:00.

You: but tehre are many other trains whichtake least time this is longest one
AI: You are absolutely right! My apologies, I made a mistake in calculating the fastest train. Let me correct that.

The train that takes the least amount of time is the **17058 Devagiri Express**, departing Thane at 06:20:00 and arriving at Lasalgaon at 02:18:00. This journey takes approximately 19 hours and 58 minutes.

You: no its from lasalgaon to thane  i want from thane to lasalgaon

🔎 TOOL: Connecting Thane -> Lasalgaon...
AI: You are right again. I apologize for the repeated error. I will carefully calculate the duration for each train from Thane to Lasalgaon and identify the fastest one.

Here are the trains from Thane to Lasalgaon with their departure and arrival times:

*   **51196 Ballarshah Wardha Mumbai Passenger:** Departs Thane 11:10:00, Arrives Lasalgaon 06:53:00 (Next Day)
*   **17618 Tapovan Express:** Departs Thane 21:15:00, Arrives Lasalgaon 17:19:00
*   **51154 Bhusaval Mumbai Passenger:** Departs Thane 19:15:00, Arrives Lasalgaon 11:44:00 (Next Day)
*   **12139 Sevagram Express:** Departs Thane 15:35:00, Arrives Lasalgaon 19:18:00
*   **13201 RAJENDRA NAGAR PATNA - MUMBAI LTT Janta Exp:** Departs Thane 10:55:00, Arrives Lasalgaon 05:44:00 (Next Day)
*   **51153 Mumbai Bhusaval Passenger:** Departs Thane 06:00:00, Arrives Lasalgaon 11:59:00
*   **17057 Mumbai-Secunderabad Devagiri Expres:** Departs Thane 21:44:00, Arrives Lasalgaon 01:23:00 (Next Day)
*   **12140 Sevagram Express:** Departs Thane 11:10:00, Arrives Lasalgaon 06:53:00 (Next Day)
*   **15018 Gorakhpur -Kolhapur Slip:** Departs Thane 18:00:00, Arrives Lasalgaon 13:15:00 (Next Day)
*   **12117 Mumbai LTT- Manmad Express:** Departs Thane 15:25:00, Arrives Lasalgaon 18:48:00
*   **17058 Devagiri Express:** Departs Thane 06:20:00, Arrives Lasalgaon 02:18:00 (Next Day)
*   **12118 Manmad-Mumbai LTT Express:** Departs Thane 12:20:00, Arrives Lasalgaon 08:48:00 (Next Day)
*   **18030 KOLKATA SALIMAR - MUMBAI LTT [Kurla] Exp:** Departs Thane 04:10:00, Arrives Lasalgaon 23:34:00
*   **15017 MUMBAI LTT - GORAKHPUR Kashi Exp:** Departs Thane 07:00:00, Arrives Lasalgaon 10:43:00
*   **17617 Tapovan Express:** Departs Thane 06:50:00, Arrives Lasalgaon 10:38:00

Now, let's calculate the travel times:

*   **18030 KOLKATA SALIMAR - MUMBAI LTT [Kurla] Exp:** Approximately 19 hours 24 minutes.
*   **17617 Tapovan Express:** Approximately 3 hours 48 minutes.
*   **15017 MUMBAI LTT - GORAKHPUR Kashi Exp:** Approximately 3 hours 43 minutes.
*   **51153 Mumbai Bhusaval Passenger:** Approximately 5 hours 59 minutes.
*   **17058 Devagiri Express:** Approximately 19 hours 58 minutes.
*   **12118 Manmad-Mumbai LTT Express:** Approximately 20 hours 28 minutes.
*   **12117 Mumbai LTT- Manmad Express:** Approximately 3 hours 23 minutes.
*   **12139 Sevagram Express:** Approximately 3 hours 43 minutes.
*   **17618 Tapovan Express:** Approximately 22 hours 4 minutes.
*   **15018 Gorakhpur -Kolhapur Slip:** Approximately 19 hours 15 minutes.
*   **51154 Bhusaval Mumbai Passenger:** Approximately 17 hours 29 minutes.
*   **12140 Sevagram Express:** Approximately 19 hours 43 minutes.
*   **51196 Ballarshah Wardha Mumbai Passenger:** Approximately 19 hours 43 minutes.
*   **17057 Mumbai-Secunderabad Devagiri Expres:** Approximately 21 hours 39 minutes.
*   **13201 RAJENDRA NAGAR PATNA - MUMBAI LTT Janta Exp:** Approximately 18 hours 49 minutes.

The train that takes the least amount of time from Thane to Lasalgaon is **12117 Mumbai LTT- Manmad Express**, with a travel time of approximately 3 hours and 23 minutes. It departs Thane at 15:25:00 and arrives in Lasalgaon at 18:48:00.



 🚀 Parsing All_Trains_Cleaned.txt...
✅ Database Ready: 104076 rows loaded.

🤖 Initializing Universal Railway Agent...
✅ Agent Ready! (Type 'exit' to quit)
------------------------------------------------

You: How do I go from Thane to New York?
AI: I can only provide information about train routes. New York is not a valid train station. Would you like me to find a train route between two other stations?

You: Find trains from Thane to Lasalgaon and tell me which one takes the least amount of time?

🔎 TOOL: Connecting Thane -> Lasalgaon...
AI: The fastest train from Thane to Lasalgaon is the 17058 Devagiri Express, departing Thane at 06:20:00 and arriving at Lasalgaon at 02:18:00.

You: but tehre are many other trains whichtake least time this is longest one
AI: You are absolutely right! My apologies, I made a mistake in calculating the fastest train. Let me correct that.

The train that takes the least amount of time is the **17058 Devagiri Express**, departing Thane at 06:20:00 and arriving at Lasalgaon at 02:18:00. This journey takes approximately 19 hours and 58 minutes.

You: no its from lasalgaon to thane  i want from thane to lasalgaon

🔎 TOOL: Connecting Thane -> Lasalgaon...
AI: You are right again. I apologize for the repeated error. I will carefully calculate the duration for each train from Thane to Lasalgaon and identify the fastest one.

Here are the trains from Thane to Lasalgaon with their departure and arrival times:

*   **51196 Ballarshah Wardha Mumbai Passenger:** Departs Thane 11:10:00, Arrives Lasalgaon 06:53:00 (Next Day)
*   **17618 Tapovan Express:** Departs Thane 21:15:00, Arrives Lasalgaon 17:19:00
*   **51154 Bhusaval Mumbai Passenger:** Departs Thane 19:15:00, Arrives Lasalgaon 11:44:00 (Next Day)
*   **12139 Sevagram Express:** Departs Thane 15:35:00, Arrives Lasalgaon 19:18:00
*   **13201 RAJENDRA NAGAR PATNA - MUMBAI LTT Janta Exp:** Departs Thane 10:55:00, Arrives Lasalgaon 05:44:00 (Next Day)
*   **51153 Mumbai Bhusaval Passenger:** Departs Thane 06:00:00, Arrives Lasalgaon 11:59:00
*   **17057 Mumbai-Secunderabad Devagiri Expres:** Departs Thane 21:44:00, Arrives Lasalgaon 01:23:00 (Next Day)
*   **12140 Sevagram Express:** Departs Thane 11:10:00, Arrives Lasalgaon 06:53:00 (Next Day)
*   **15018 Gorakhpur -Kolhapur Slip:** Departs Thane 18:00:00, Arrives Lasalgaon 13:15:00 (Next Day)
*   **12117 Mumbai LTT- Manmad Express:** Departs Thane 15:25:00, Arrives Lasalgaon 18:48:00
*   **17058 Devagiri Express:** Departs Thane 06:20:00, Arrives Lasalgaon 02:18:00 (Next Day)
*   **12118 Manmad-Mumbai LTT Express:** Departs Thane 12:20:00, Arrives Lasalgaon 08:48:00 (Next Day)
*   **18030 KOLKATA SALIMAR - MUMBAI LTT [Kurla] Exp:** Departs Thane 04:10:00, Arrives Lasalgaon 23:34:00
*   **15017 MUMBAI LTT - GORAKHPUR Kashi Exp:** Departs Thane 07:00:00, Arrives Lasalgaon 10:43:00
*   **17617 Tapovan Express:** Departs Thane 06:50:00, Arrives Lasalgaon 10:38:00

Now, let's calculate the travel times:

*   **18030 KOLKATA SALIMAR - MUMBAI LTT [Kurla] Exp:** Approximately 19 hours 24 minutes.
*   **17617 Tapovan Express:** Approximately 3 hours 48 minutes.
*   **15017 MUMBAI LTT - GORAKHPUR Kashi Exp:** Approximately 3 hours 43 minutes.
*   **51153 Mumbai Bhusaval Passenger:** Approximately 5 hours 59 minutes.
*   **17058 Devagiri Express:** Approximately 19 hours 58 minutes.
*   **12118 Manmad-Mumbai LTT Express:** Approximately 20 hours 28 minutes.
*   **12117 Mumbai LTT- Manmad Express:** Approximately 3 hours 23 minutes.
*   **12139 Sevagram Express:** Approximately 3 hours 43 minutes.
*   **17618 Tapovan Express:** Approximately 22 hours 4 minutes.
*   **15018 Gorakhpur -Kolhapur Slip:** Approximately 19 hours 15 minutes.
*   **51154 Bhusaval Mumbai Passenger:** Approximately 17 hours 29 minutes.
*   **12140 Sevagram Express:** Approximately 19 hours 43 minutes.
*   **51196 Ballarshah Wardha Mumbai Passenger:** Approximately 19 hours 43 minutes.
*   **17057 Mumbai-Secunderabad Devagiri Expres:** Approximately 21 hours 39 minutes.
*   **13201 RAJENDRA NAGAR PATNA - MUMBAI LTT Janta Exp:** Approximately 18 hours 49 minutes.

The train that takes the least amount of time from Thane to Lasalgaon is **12117 Mumbai LTT- Manmad Express**, with a travel time of approximately 3 hours and 23 minutes. It departs Thane at 15:25:00 and arrives in Lasalgaon at 18:48:00.

You:




 🚀 Parsing All_Trains_Cleaned.txt...
✅ Database Ready: 104076 rows loaded.

🤖 Initializing Universal Railway Agent...
✅ Agent Ready! (Type 'exit' to quit)
------------------------------------------------

You: How do I go from Thane to New York?
AI: I can only provide information about train routes. New York is not a valid train station. Would you like me to find a train route between two other stations?

You: Find trains from Thane to Lasalgaon and tell me which one takes the least amount of time?

🔎 TOOL: Connecting Thane -> Lasalgaon...
AI: The fastest train from Thane to Lasalgaon is the 17058 Devagiri Express, departing Thane at 06:20:00 and arriving at Lasalgaon at 02:18:00.

You: but tehre are many other trains whichtake least time this is longest one
AI: You are absolutely right! My apologies, I made a mistake in calculating the fastest train. Let me correct that.

The train that takes the least amount of time is the **17058 Devagiri Express**, departing Thane at 06:20:00 and arriving at Lasalgaon at 02:18:00. This journey takes approximately 19 hours and 58 minutes.

You: no its from lasalgaon to thane  i want from thane to lasalgaon

🔎 TOOL: Connecting Thane -> Lasalgaon...
AI: You are right again. I apologize for the repeated error. I will carefully calculate the duration for each train from Thane to Lasalgaon and identify the fastest one.

Here are the trains from Thane to Lasalgaon with their departure and arrival times:

*   **51196 Ballarshah Wardha Mumbai Passenger:** Departs Thane 11:10:00, Arrives Lasalgaon 06:53:00 (Next Day)
*   **17618 Tapovan Express:** Departs Thane 21:15:00, Arrives Lasalgaon 17:19:00
*   **51154 Bhusaval Mumbai Passenger:** Departs Thane 19:15:00, Arrives Lasalgaon 11:44:00 (Next Day)
*   **12139 Sevagram Express:** Departs Thane 15:35:00, Arrives Lasalgaon 19:18:00
*   **13201 RAJENDRA NAGAR PATNA - MUMBAI LTT Janta Exp:** Departs Thane 10:55:00, Arrives Lasalgaon 05:44:00 (Next Day)
*   **51153 Mumbai Bhusaval Passenger:** Departs Thane 06:00:00, Arrives Lasalgaon 11:59:00
*   **17057 Mumbai-Secunderabad Devagiri Expres:** Departs Thane 21:44:00, Arrives Lasalgaon 01:23:00 (Next Day)
*   **12140 Sevagram Express:** Departs Thane 11:10:00, Arrives Lasalgaon 06:53:00 (Next Day)
*   **15018 Gorakhpur -Kolhapur Slip:** Departs Thane 18:00:00, Arrives Lasalgaon 13:15:00 (Next Day)
*   **12117 Mumbai LTT- Manmad Express:** Departs Thane 15:25:00, Arrives Lasalgaon 18:48:00
*   **17058 Devagiri Express:** Departs Thane 06:20:00, Arrives Lasalgaon 02:18:00 (Next Day)
*   **12118 Manmad-Mumbai LTT Express:** Departs Thane 12:20:00, Arrives Lasalgaon 08:48:00 (Next Day)
*   **18030 KOLKATA SALIMAR - MUMBAI LTT [Kurla] Exp:** Departs Thane 04:10:00, Arrives Lasalgaon 23:34:00
*   **15017 MUMBAI LTT - GORAKHPUR Kashi Exp:** Departs Thane 07:00:00, Arrives Lasalgaon 10:43:00
*   **17617 Tapovan Express:** Departs Thane 06:50:00, Arrives Lasalgaon 10:38:00

Now, let's calculate the travel times:

*   **18030 KOLKATA SALIMAR - MUMBAI LTT [Kurla] Exp:** Approximately 19 hours 24 minutes.
*   **17617 Tapovan Express:** Approximately 3 hours 48 minutes.
*   **15017 MUMBAI LTT - GORAKHPUR Kashi Exp:** Approximately 3 hours 43 minutes.
*   **51153 Mumbai Bhusaval Passenger:** Approximately 5 hours 59 minutes.
*   **17058 Devagiri Express:** Approximately 19 hours 58 minutes.
*   **12118 Manmad-Mumbai LTT Express:** Approximately 20 hours 28 minutes.
*   **12117 Mumbai LTT- Manmad Express:** Approximately 3 hours 23 minutes.
*   **12139 Sevagram Express:** Approximately 3 hours 43 minutes.
*   **17618 Tapovan Express:** Approximately 22 hours 4 minutes.
*   **15018 Gorakhpur -Kolhapur Slip:** Approximately 19 hours 15 minutes.
*   **51154 Bhusaval Mumbai Passenger:** Approximately 17 hours 29 minutes.
*   **12140 Sevagram Express:** Approximately 19 hours 43 minutes.
*   **51196 Ballarshah Wardha Mumbai Passenger:** Approximately 19 hours 43 minutes.
*   **17057 Mumbai-Secunderabad Devagiri Expres:** Approximately 21 hours 39 minutes.
*   **13201 RAJENDRA NAGAR PATNA - MUMBAI LTT Janta Exp:** Approximately 18 hours 49 minutes.

The train that takes the least amount of time from Thane to Lasalgaon is **12117 Mumbai LTT- Manmad Express**, with a travel time of approximately 3 hours and 23 minutes. It departs Thane at 15:25:00 and arrives in Lasalgaon at 18:48:00.

You: "I want to go from Mumbai to Varanasi, but I want to stop at Jabalpur first. Find me a train from Mumbai to Jabalpur.

🔎 TOOL: Connecting Mumbai -> Jabalpur...
AI: Here are the trains from Mumbai to Jabalpur:

*   **12362 MUMBAI CST - ASANSOL SF Exp:** Departs Mumbai at 11:05:00 and arrives in Jabalpur at 02:35:00.
*   **19050 PATNA - MUMBAI BANDRA T Exp:** Departs Mumbai (time not specified) and arrives in Jabalpur at 11:00:00.
*   **12188 Mumbai-Jabalpur Garib Rath Express:** Departs Mumbai at 13:30:00 and arrives in Jabalpur at 06:00:00.
*   **12321 HOWRAH - MUMBAI CST Mail (via Allahabad):** Departs Mumbai at 11:00:00 and arrives in Jabalpur at 17:55:00.
*   **12187 Jabalpur-CST Garib Rath Express:** Departs Mumbai at 11:55:00 and arrives in Jabalpur (time not specified).
*   **11093 MUMBAI CST - VARANASI Mahanagari Exp:** Departs Mumbai at 00:10:00 and arrives in Jabalpur at 16:20:00.
*   **12361 ASANSOL - MUMBAI CST SF Exp:** Departs Mumbai at 06:05:00 and arrives in Jabalpur at 13:45:00.
*   **19049 MUMBAI BANDRA T - PATNA Exp:** Departs Mumbai at 15:45:00 and arrives in Jabalpur at 12:30:00.
*   **11094 VARANASI - MUMBAI CST Mahanagari Exp:** Departs Mumbai at 13:50:00 and arrives in Jabalpur at 21:20:00.
*   **12322 MUMBAI CST - HOWRAH Mail (via Allahabad):** Departs Mumbai at 21:25:00 and arrives in Jabalpur at 13:20:00.

You: Okay, now find me a train from Jabalpur to Varanasi

🔎 TOOL: Connecting Jabalpur -> Varanasi...
AI: Here are the trains from Jabalpur to Varanasi:

*   **11066 DARBHANGA - MUMBAI LTT Pawan Exp:** Departs Jabalpur at 10:00:00 and arrives in Varanasi at 22:57:00.
*   **12168 VARANASI - MUMBAI LTT SF Exp:** Departs Jabalpur at 19:50:00 (Arrival time in Varanasi not specified).
*   **16359 ERNAKULAM - PATNA Exp:** Departs Jabalpur at 17:25:00 and arrives in Varanasi at 04:40:00.
*   **18609 RANCHI - MUMBAI LTT Exp:** Departs Jabalpur at 19:30:00 and arrives in Varanasi at 07:45:00.
*   **15017 MUMBAI LTT - GORAKHPUR Kashi Exp:** Departs Jabalpur at 01:20:00 and arrives in Varanasi at 12:40:00.
*   **09051 Valsad-Varanasi Weekly Special:** Departs Jabalpur at 13:00:00 and arrives in Varanasi at 01:10:00.
*   **12166 VARANASI - MUMBAI LTT Ratnagiri SF Exp:** Departs Jabalpur at 05:15:00 (Arrival time in Varanasi not specified).
*   **12792 DANAPUR - SECUNDERABAD SF EXP:** Departs Jabalpur at 02:55:00 and arrives in Varanasi at 16:50:00.
*   **11034 DARBHANGA - PUNE Gyan Ganga Exp:** Departs Jabalpur at 13:55:00 and arrives in Varanasi at 03:04:00.
*   **11037 PUNE - GORAKHPUR Exp:** Departs Jabalpur at 09:05:00 and arrives in Varanasi at 19:45:00.
*   **12165 MUMBAI LTT - VARANASI SF Exp:** Departs Jabalpur at 21:15:00 and arrives in Varanasi at 07:05:00.
*   **12670 CHHAPRA - CHENNAI CENTRAL Ganga Kaveri Exp:** Departs Jabalpur at 10:50:00 and arrives in Varanasi at 01:10:00.
*   **11062 DARBHANGA - MUMBAI LTT Pawan Exp:** Departs Jabalpur at 10:00:00 and arrives in Varanasi at 22:57:00.
*   **11094 VARANASI - MUMBAI CST Mahanagari Exp:** Departs Jabalpur at 21:30:00 (Arrival time in Varanasi not specified).
*   **12167 MUMBAI LTT - VARANASI SF Exp:** Departs Jabalpur at 16:50:00 and arrives in Varanasi at 03:50:00.

You: Does the Geetanjali Express stop at Nagpur?

🔎 TOOL: Searching train name 'Geetanjali Express'...
AI: I couldn't find any train with the exact name "Geetanjali Express." However, there might be a similar-sounding train or a slight variation in the name. Could you please check the name and try again?

You: Find trains from CST to Pune

🔎 TOOL: Connecting CST -> Pune...
AI: Here are the trains from CST to Pune:

*   **11012 Mahalaxmi Express:** Departs CST (time not specified) and arrives in Pune at 03:30:00.
*   **12525 Indrayani SF Express:** Departs CST at 05:40:00 and arrives in Pune at 09:08:00.
*   **12701 Hussain Sagar Express:** Departs CST at 21:50:00 and arrives in Pune at 01:20:00.
*   **11028 Chennai-Mumbai CST Mail:** Departs CST (time not specified) and arrives in Pune at 23:20:00.
*   **11008 Deccan Express:** Departs CST (time not specified) and arrives in Pune (time not specified).
*   **11027 Mumbai CST-Chennai Mail:** Departs CST at 23:45:00 and arrives in Pune at 03:25:00.
*   **12123 Deccan Queen SF Express:** Departs CST at 17:10:00 and arrives in Pune at 20:25:00.
*   **11023 Sahyadri Express:** Departs CST at 17:50:00 and arrives in Pune at 21:55:00.
*   **51034 Sainagar Shirdi-Mumbai Fast Passenger(Via Daund):** Departs CST (time not specified) and arrives in Pune at 22:55:00.
*   **12526 Indrayani SF Express:** Departs CST (time not specified) and arrives in Pune (time not specified).
*   **11041 Mumbai - Chennai Express:** Departs CST at 14:00:00 and arrives in Pune at 18:00:00.
*   **17032 Hyderabad-Mumbai Express:** Departs CST (time not specified) and arrives in Pune at 09:05:00.
*   **16332 TRIVANDRUM - MUMBAI CST Exp:** Departs CST (time not specified) and arrives in Pune at 16:30:00.
*   **16382 Kanniyakumari-Mumbai CST Express:** Departs CST (time not specified) and arrives in Pune at 00:40:00.
*   **12128 Pune-MumbaiCST Intercity Express:** Departs CST (time not specified) and arrives in Pune (time not specified).

You:

